File: extract-linkedin-premium-users-from-linkedin-groups/agentic/langGraph/tools/linkedin/fetch-linkedIn-group-members-tool.ts
/// <reference types="node" />
import { tool } from "@langchain/core/tools";
import { z } from "zod";

const CONNECTSAFELY_API_TOKEN = process.env.CONNECTSAFELY_API_TOKEN || "";

/**
 * Tool 1: Fetch LinkedIn group members with pagination
 */
export const fetchLinkedInGroupMembersTool = tool(
  async ({ groupId, start, count }) => {
    const actualStart = start ?? 0;
    const actualCount = count ?? 50;
    
    const res = await fetch(
      "https://api.connectsafely.ai/linkedin/groups/members",
      {
        method: "POST",
        headers: {
          Authorization: `Bearer ${CONNECTSAFELY_API_TOKEN}`,
          "Content-Type": "application/json",
        },
        body: JSON.stringify({ groupId, start: actualStart, count: actualCount }),
      }
    );

    if (!res.ok) {
      throw new Error(`Failed to fetch members: ${res.statusText}`);
    }

    const data = await res.json() as { members?: any[]; hasMore?: boolean };
    
    return JSON.stringify({
      members: data.members || [],
      hasMore: data.hasMore || false,
      fetched: (data.members || []).length,
    });
  },
  {
    name: "fetch-linkedin-group-members",
    description: "Fetch a single paginated batch of LinkedIn group members. Use for low-level control.",
    schema: z.object({
      groupId: z.string().describe("The LinkedIn group ID"),
      start: z.number().optional().describe("Starting offset for pagination"),
      count: z.number().optional().describe("Number of members to fetch (max 100)"),
    }),
  }
);

File: extract-linkedin-premium-users-from-linkedin-groups/agentic/agno/tools/linkedin/fetch_linkedIn_group_members_tool.py
import os
from dotenv import load_dotenv
import httpx
import json
from typing import Optional

load_dotenv()

def fetch_linkedin_group_members(group_id: str, start: int = 0, count: int = 50) -> str:
    """
    Fetch a single paginated batch of LinkedIn group members. Use for low-level control.
    
    Args:
        group_id (str): The LinkedIn group ID.
        start (int): Starting offset for pagination (default 0).
        count (int): Number of members to fetch, max 100 (default 50).
    """
    api_token = os.getenv("CONNECTSAFELY_API_TOKEN", "")
    url = "https://api.connectsafely.ai/linkedin/groups/members"
    headers = {
        "Authorization": f"Bearer {api_token}",
        "Content-Type": "application/json",
    }
    
    payload = {
        "groupId": group_id,
        "start": start,
        "count": count
    }

    try:
        with httpx.Client(timeout=30.0) as client:
            response = client.post(url, headers=headers, json=payload)
            response.raise_for_status()
            data = response.json()
            
            # Formating the return to match your TS tool's output structure
            result = {
                "members": data.get("members", []),
                "hasMore": data.get("hasMore", False),
                "fetched": len(data.get("members", []))
            }
            return json.dumps(result)
            
    except Exception as e:
        return f"Error fetching batch: {str(e)}"